### PR TITLE
Use sparse.COO for our multiplication table

### PR DESCRIPTION
**Pull request by eric-wieser**
An alternative to #133, that shares the first two commits

I plan to squash these commits when I merge, the history isn't that interesting

### CODE REVIEW

File: clifford/__init__.py [hunk 1]
```diff
@@ -139,73 +142,70 @@ def construct_tables(
...
                       grades_a=None, grades_b=None, filter_mask=None):
     '''
     Returns a function that implements the mult_table on two input multivectors
     '''
     if product_mask is not None:
-        k_list = k_list[product_mask]
-        l_list = l_list[product_mask]
-        m_list = m_list[product_mask]
-        mult_table_vals = mult_table_vals[product_mask]
+        mt = sparse.where(product_mask, mt, 0)
```
Comment (eric-wieser) [inline clifford/__init__.py]: Strictly this is a little less efficient, but it has the advantage of working with dense product masks too. At any rate, this is only a startup cost.

Comment (hugohadfield) [reply clifford/__init__.py]: How does this start up impact effect large-ish d algebras eg Cl(8)?

Comment (eric-wieser) [reply clifford/__init__.py]: Before:
```
2.16 s ± 47.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
After:
```
In [2]: %timeit clifford.Cl(8)
1.34 s ± 64.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

???

Comment (hugohadfield) [reply clifford/__init__.py]: Nice

**Comment by eric-wieser** [comment]
A little worried that `test_rotor_between_non_overlapping_spheres` is failing on both travis conda runs

**Comment by eric-wieser** [comment]
I restarted the travis and a different test failed - looks like that was just coincidence

**Pull request merged**